### PR TITLE
[builder] Support for --skip-new-go-module

### PR DESCRIPTION
**Pull request by kristinapathak**
A continuation of https://github.com/open-telemetry/opentelemetry-collector/pull/9253 and https://github.com/open-telemetry/opentelemetry-collector/pull/9631

Description: Adds a `--skip-new-go-module` flag to the OTC builder. This enables users working in an existing go module environment (say, a "monorepo") to update the module they have, vs forcing the use of a new module.

With the new support inside an existing Go module, a collector main package can be generated using a go:generate directive. For example, in the directory where I want my collector built, the file generate.go has this line:

//go:generate builder --skip-new-go-module --skip-compilation --strict-versioning --config=./build-config.yaml
In the same directory, the build-config.yaml describes the collector to build. The builder generates the other files in the same directory. At this point, normal Go workflows can be used to update indirect dependencies.

Link to tracking Issue: https://github.com/open-telemetry/opentelemetry-collector/issues/9252

Testing: Will add unit tests in the next few days.

Documentation: Yes.

### CODE REVIEW

File: cmd/builder/internal/builder/config.go
Comment (mx-psi) [inline cmd/builder/internal/builder/config.go:248]: You can just use the `%q` verb for quoting
```suggestion
			return fmt.Errorf("%w cannot modify mod.path %q combined with --skip-new-go-module; please modify the enclosing go.mod file directly", ErrIncompatibleConfigurationValues, mod.Path)
```

File: cmd/builder/README.md
Comment (mx-psi) [inline cmd/builder/README.md:159]: I would keep it a bit simpler
```suggestion
You can optionally skip creating a new `go.mod` file.
When the `--skip-new-go-module` command-line flag is supplied,
```

It would also be nice to motivate this: when would an end-user be interested in using `--skip-new-go-module`?

File: cmd/builder/README.md
Comment (mx-psi) [inline cmd/builder/README.md:173]: nit: make it a bit more explicit

```suggestion
`--skip-new-go-module` is incompatible with `replaces`, `excludes`,
and the component `path` override.  For each of these features, users
are expected to modify the enclosing `go.mod` directly.
```

File: cmd/builder/internal/builder/main.go
Comment (mx-psi) [inline cmd/builder/internal/builder/main.go:274]:
```suggestion
	mod, ver, found := strings.Cut(modspec, " ")
	if !found {
		return fmt.Errorf("ill-formatted modspec %q: missing space separator", modspec)
	}
```

File: cmd/builder/internal/builder/main.go
Comment (mx-psi) [inline cmd/builder/internal/builder/main.go:297]: I am not sure which one would be better, but we also have the option of doing `go mod edit -require=example.org@v1.2.0`. Why is `go get` preferred here?

File: cmd/builder/internal/command.go
Comment (mx-psi) [inline cmd/builder/internal/command.go:88]: I am not a native speaker, but this sounds better?
```suggestion
	cmd.Flags().BoolVar(&cfg.SkipNewGoModule, skipNewGoModuleFlag, false, "Whether builder should skip generating a new go.mod file, using the enclosing Go module instead (default false)")
```

File: cmd/builder/internal/builder/main.go
Comment (kristinapathak) [reply cmd/builder/internal/builder/main.go:300]: From what I can tell, the difference is that `go get` installs and builds the packages needed whereas `go mod edit` doesn't.  Given that we run `go mod tidy` after this in `GetModules()`, I'll change this to use `go mod edit`.

Comment (kristinapathak) [reply cmd/builder/internal/builder/main.go:300]: One thing to [note](https://go.dev/ref/mod#go-mod-edit) about using `go mod edit -require`:
>These flags are mainly for tools that understand the module graph. Users should prefer go get path@version or go get path@none, which make other go.mod adjustments as needed to satisfy constraints imposed by other modules. 

I'm not sure what these other adjustments are or if we need them here? I would expect if we change the go.mod into some incompatible graph that `go mod tidy` will fail after this.

Comment (mx-psi) [reply cmd/builder/internal/builder/main.go]: Thinking more about this, I think I would prefer using `go mod edit -require` at first. The section you quoted sounds to me as "`go get` also does `go mod tidy`, which you probably want", and since, as you point out, we already `go mod tidy` after this, I would prefer to keep things simple

**Review by mx-psi — COMMENTED** [review]
Is there any chance to increase testing coverage here? Some error cases would be harder to reach, but there are other parts that should be doable

**Comment by kristinapathak** [comment]
@mx-psi, sorry for the delay.  I hit this issue when adding back tests and then got pulled into other things:
```
failed to update go.mod: go subcommand failed with args '[mod tidy -compat=1.21]': exit status 1, error message: go: finding module for package go.opentelemetry.io/collector/internal/testdata
        	            	go: finding module for package go.opentelemetry.io/collector/confmap/converter/expandconverter
        	            	go: found go.opentelemetry.io/collector/confmap/converter/expandconverter in go.opentelemetry.io/collector/confmap/converter/expandconverter v0.101.0
        	            	go: finding module for package go.opentelemetry.io/collector/internal/testdata
        	            	go: go.opentelemetry.io/collector/cmd/builder/internal/tester/output imports
        	            		go.opentelemetry.io/collector/connector tested by
        	            		go.opentelemetry.io/collector/connector.test imports
        	            		go.opentelemetry.io/collector/internal/testdata: module go.opentelemetry.io/collector@latest found (v0.101.0), but does not contain package go.opentelemetry.io/collector/internal/testdata

```

I'll try to fix this and increase coverage more this week.

**Comment by mx-psi** [comment]
@kristinapathak feel free to ping me when tests are passing :)

File: cmd/builder/internal/builder/main.go
Comment (kristinapathak) [inline cmd/builder/internal/builder/main.go:155]: `go mod tidy` is run a few lines above and downloads the needed dependencies, so `go mod download` isn't needed. No dependencies are updated after the tidy call.

File: cmd/builder/internal/builder/main.go
Comment (kristinapathak) [inline cmd/builder/internal/builder/main.go:285]: I couldn't figure out how to get `go mod edit` to fail 😞 everything I tried caused other failures...

**Review by mx-psi — APPROVED** [review]

File: cmd/builder/internal/builder/main.go
Comment (mx-psi) [reply cmd/builder/internal/builder/main.go:154]: Thanks, this comment was useful!

**Comment by mx-psi** [comment]
CI failures are due to #10661, will merge this PR after merging that one

**Comment by mx-psi** [comment]
```
--- FAIL: TestGetModules (36.26s)
    --- FAIL: TestGetModules/Exporter_Not_in_Gomod (3.57s)
        main_test.go:633: 
            	Error Trace:	D:/a/opentelemetry-collector/opentelemetry-collector/cmd/builder/internal/builder/main_test.go:633
            	Error:      	Received unexpected error:
            	            	failed to compile the OpenTelemetry Collector distribution: go subcommand failed with args '[build -trimpath -o  -ldflags=-s -w]': exit status 1, error message: # go.opentelemetry.io/collector/config/configgrpc
Error:             	            	C:\Users\runneradmin\go\pkg\mod\go.opentelemetry.io\collector\config\configgrpc@v0.102.1\configgrpc.go:274:62: not enough arguments in call to gcs.Auth.GetClientAuthenticator
            	            		have (map[component.ID]component.Component)
            	            		want (context.Context, map[component.ID]component.Component)
```

**Comment by kristinapathak** [comment]
@mx-psi, can this be merged?

**Comment by mx-psi** [comment]
@kristinapathak Sorry, should have said something explicitly, CI is not passing so I can't merge (it's a required CI so I literally can't click the button :smile:). This error is still happening https://github.com/open-telemetry/opentelemetry-collector/pull/10098#issuecomment-2236632918

**Comment by mx-psi** [comment]
Looks like CI is still having trouble

**Comment by kristinapathak** [comment]
@mx-psi, tests are back to passing 🙂

**Comment by mx-psi** [comment]
Looks like there is one last test that is timing out on Windows:

```
panic: test timed out after 4m0s
running tests:
	TestGetModules (17s)
	TestGetModules/Exporter_Not_in_Gomod (3s)
```

I am fine with increasing the timeout value. Could you go ahead and do that?

**Comment by mx-psi** [comment]
Looks like bumping the timeout revealed an error:

```
	Error:      	Error "component dependency not found in go mod file: 'go.opentelemetry.io/collector/confmap/provider/envprovider'. Use --skip-strict-versioning to temporarily disable this check. This flag will be removed in a future minor version" does not contain "failed to generate source file"
```

**Comment by mx-psi** [comment]
:thinking: 

Another error

```
--- FAIL: TestGenerateAndCompile (74.27s)
    --- FAIL: TestGenerateAndCompile/No_Dir_Permissions (2.51s)
        main_test.go:441: 
            	Error Trace:	D:/a/opentelemetry-collector/opentelemetry-collector/cmd/builder/internal/builder/main_test.go:441
            	Error:      	An error is expected but got nil.
            	Test:       	TestGenerateAndCompile/No_Dir_Permissions
```

**Comment by mx-psi** [comment]
:tada: :tada:

**Pull request merged**
